src/metrics.rs

```diff
@@ -29,118 +28,75 @@ pub use export_to_sheets::ExportToSheets;
 pub use list_repos::ListReposForOrg;
 pub use print::Print;
 
-#[throws]
-async fn all_repos(org: &octocrab::orgs::OrgHandler<'_>) -> Vec<octocrab::models::Repository> {
-    util::accumulate_pages(|page| org.list_repos().page(page).send()).await?
-}
-
-// ==================== GQL Query Structures =========================
+#[derive(GraphQLQuery)]
+#[graphql(
+    schema_path = "gql/schema.docs.graphql",
+    query_path = "gql/query_search.graphql",
+    response_derives = "Serialize,Debug"
+)]
+pub struct QuerySearch;
+
+#[derive(GraphQLQuery)]
+#[graphql(
+    schema_path = "gql/schema.docs.graphql",
+    query_path = "gql/organization_repos.graphql",
+    response_derives = "Serialize,Debug"
+)]
+pub struct OrgRepos;
 
-#[derive(Deserialize, Debug)]
-pub struct Response<T> {
-    data: T,
-    errors: Option<Vec<serde_json::Value>>,
-}
-
-#[derive(Deserialize, Debug)]
-pub struct PageInfo {
-    #[serde(rename = "hasNextPage")]
-    has_next_page: bool,
-    #[serde(rename = "endCursor")]
-    end_cursor: String,
-}
-
-#[derive(Deserialize, Debug)]
-struct AllReposData {
-    organization: Organization,
-}
-
-#[derive(Deserialize, Debug)]
-struct Organization {
-    repositories: Repositories,
-}
-
-#[derive(Deserialize, Debug)]
-struct Repositories {
-    edges: Vec<Node>,
-    #[serde(rename = "pageInfo")]
-    page_info: PageInfo,
-}
-
-#[derive(Deserialize, Debug)]
-struct Node {
-    node: RepoNode,
+#[async_trait]
+pub trait GQL {
+    async fn execute<Q: GraphQLQuery + Send>(self) -> octocrab::Result<Response<Q::ResponseData>>;
 }
 
-#[derive(Deserialize, Debug)]
-struct RepoNode {
-    name: String,
+#[async_trait]
+impl<V> GQL for QueryBody<V>
+where
+    V: Serialize + Send + Sync,
+{
+    async fn execute<Q: GraphQLQuery + Send>(self) -> octocrab::Result<Response<Q::ResponseData>> {
```

Comment:
Hmm, it's annoying that you need the `Q` parameter here! Now I'm thinking we should refactor this once more (sorry), to something like this:

```rust
impl<Q> GQL for Q
where
    Q: GraphQLQuery,
{
    fn execute(variables: Self::Variables) -> octocrab::Result<Response<Self::ResponseData>> {
        let body = Self::build(variables);
        Ok(octocrab::instance().post("graphql", Some(&body)).await?)
    }
}
```

Reply:
how would you denote this in the trait itself??

Would I need `pub trait GQL<Q>`, essentially? I'm not entirely sure how to move that type parameter that was attached to `execute`

Reply:
jk I think I figured it out